### PR TITLE
Dynamic adding and removing of rtr_mgr_groups

### PR DESCRIPTION
**Pull request by reuteran**
This uses a linked list instead of an array to store rtr_mgr_groups. It also adds two new functions for the user: rtr_mgr_group_add and rtr_mgr_group_remove. These allow users to add and remove groups while rtr_mgr is running. 

Right now this uses a tommy_list out of convenience. This is a generic doubly linked list. 
This means there are now calls to tommy_* functions in rtr_mgr.c 
Do we want to write our own linked list? Or some kind of wrapper for tommy_list?

Since I am in the middle of writing my masters thesis, it would be good if @ColinBS and @mroethke could pick this up and add necessary changes.  This should include more extensive tests, such as adding/removing groups while rtr_mgr is running and checking whether the changes work as they should there as well.

### CODE REVIEW

File: rtrlib/rtr_mgr.h [hunk 1]
```diff
@@ -67,28 +68,37 @@ struct rtr_mgr_group {
 	enum rtr_mgr_status status;
 };
 
+//TODO Find a nicer way todo a linked list (without writing our own)
```
Comment (fho) [inline rtrlib/rtr_mgr.h]: I recommend to use the linked list from the linux kernel https://github.com/torvalds/linux/blob/master/include/linux/list.h.
Some small modifications needs to be done to make it work without the other kernel sources  but then it's awesome :-)

**Comment by waehlisch** [comment]
@fho, do you have time to give this PR a review? Would be nice to merge this soon, as this is required to finish the FRR (fork of Quagga) extension.

**Comment by fho** [comment]
@waehlisch Yes, I'll do a review latest on the weekend

File: rtrlib/rtr_mgr.h [hunk 2]
```diff
@@ -67,28 +68,37 @@ struct rtr_mgr_group {
...
+
 typedef void (*rtr_mgr_status_fp)(const struct rtr_mgr_group *,
 				  enum rtr_mgr_status,
 				  const struct rtr_socket *,
 				  void *);
 
+//TODO Add refresh, expire, and retry intervals to config for easier access.
 struct rtr_mgr_config {
-	struct rtr_mgr_group *groups;
+	tommy_list groups;
```
Comment (fho) [inline rtrlib/rtr_mgr.h]: It would be great to hide the actual list implementation from the user. This would allow to change freely the data structure in the background without breaking the API.
I also don't like the idea much that the user has additional to know the tommy_list API to use the rtrlib.

The `struct rtr_mgr_group` should be a opaque data structure that is only accessible via helper functions, something like:
```
struct rtr_mgr_group
struct rtr_mgr_group *rtr_group_alloc();
void rtr_group_free(struct rtr_mgr_group *);
int rtr_mgr_group_append(struct rtr_mgr_group *list, struct rtr_mgr_group *new);
void rtr_mgr_group_remove(struct rtr_mgr_group *grp);


void rtr_mgr_group_add_socket(struct rtr_mgr group *, struct rtr_socket *);
void rtr_mgr_group_set_pref(struct rtr_mgr group *, uint8_t pref);
...
```
Refactoring that API might be a separate topic. The changes in the this pull request are consistent with the previous API but it makes it even more complex :-)

File: rtrlib/rtr_mgr.c [hunk 3]
```diff
@@ -58,21 +59,24 @@ static int rtr_mgr_start_sockets(struct rtr_mgr_group *group)
 	return RTR_SUCCESS;
 }
 
-int rtr_mgr_find_group(const struct rtr_mgr_config *config,
-		       const struct rtr_socket *sock,
-		       unsigned int *ind)
+static struct rtr_mgr_group *rtr_mgr_find_group(struct rtr_mgr_config *config,
+						const struct rtr_socket *sock)
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: 1 space for indentation missing :-)

File: rtrlib/rtr_mgr.c [hunk 4]
```diff
@@ -58,21 +59,24 @@ static int rtr_mgr_start_sockets(struct rtr_mgr_group *group)
...
-			}
+	tommy_node *node;
+	struct rtr_mgr_group_node *group_node;
+
+	node  = tommy_list_head(&config->groups);
+
+	while (node) {
+		group_node = node->data;
+		for (unsigned int j = 0; j < group_node->group->sockets_len;
+									 j++) {
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: why is `j++` on a separate line?

File: rtrlib/rtr_mgr.c [hunk 5]
```diff
@@ -58,21 +59,24 @@ static int rtr_mgr_start_sockets(struct rtr_mgr_group *group)
...
-				return RTR_SUCCESS;
-			}
+	tommy_node *node;
+	struct rtr_mgr_group_node *group_node;
+
+	node  = tommy_list_head(&config->groups);
+
+	while (node) {
+		group_node = node->data;
+		for (unsigned int j = 0; j < group_node->group->sockets_len;
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: `int j` is sufficient and shorter

File: rtrlib/rtr_mgr.c [hunk 6]
```diff
@@ -431,17 +507,148 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+	}
+}
+
+int rtr_mgr_add_group(struct rtr_mgr_config *config,
+		      const struct rtr_mgr_group *group)
+{
+	unsigned int refresh_iv = 3600;
+	unsigned int retry_iv = 600;
+	unsigned int expire_iv = 7200;
+	int err_code = RTR_ERROR;
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: instead of int this could be the corresponding enum type

File: tests/test_live_validation.c [hunk 7]
```diff
@@ -116,7 +117,7 @@ int main(void)
 
 	rtr_mgr_stop(conf);
 	rtr_mgr_free(conf);
-	free(groups[0].sockets);
+	//free(groups[0].sockets);
```
Comment (fho) [inline tests/test_live_validation.c]: either remove the comment or uncomment the line

File: rtrlib/rtr_mgr.c [hunk 8]
```diff
@@ -431,17 +507,148 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+
+	new_group_node->group = new_group;
+	tommy_list_insert_tail(&config->groups, &new_group_node->node,
+			       new_group_node);
+	config->len++;
+
+	tommy_list_sort(&config->groups, &rtr_mgr_config_cmp);
+	pthread_mutex_unlock(&config->mutex);
+
+	struct rtr_mgr_group *best_group = rtr_mgr_get_first_group(config);
```
Comment (nbars) [inline rtrlib/rtr_mgr.c]: This pointer might not point to a valid memory region (in the following instructions), as we hold no lock and the group could be freed in the meantime.

File: rtrlib/rtr_mgr.c [hunk 9]
```diff
@@ -90,135 +94,155 @@ bool rtr_mgr_config_status_is_synced(const struct rtr_mgr_group *group)
...
-	/* groups are sorted by preference */
-	for (int i = 0; i < config->len; i++) {
-		if ((i != my_group_idx) &&
-		    (config->groups[i].status == RTR_MGR_CLOSED))
-			return &config->groups[i];
+	tommy_node *node = tommy_list_head(&config->groups);
+	struct rtr_mgr_group_node *group_node;
+	struct rtr_mgr_group *group;
+
+	while (node) {
```
Comment (nbars) [inline rtrlib/rtr_mgr.c]: indention is missing here :-)

File: rtrlib/rtr_mgr.c [hunk 10]
```diff
@@ -431,17 +507,148 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
 void rtr_mgr_stop(struct rtr_mgr_config *config)
 {
+	tommy_node *node = tommy_list_head(&config->groups);
+	struct rtr_mgr_group_node *group_node;
+
 	MGR_DBG1("rtr_mgr_stop()");
-	for (unsigned int i = 0; i < config->len; i++) {
-		for (unsigned int j = 0; j < config->groups[i].sockets_len; j++)
-			rtr_stop(config->groups[i].sockets[j]);
+	while (node) {
```
Comment (nbars) [inline rtrlib/rtr_mgr.c]: No lock held while iterating over the list.

File: rtrlib/rtr_mgr.c (hunk 6 above)
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: In #98 there are default values for these intervals. Could they serve the purpose for this problem?

```
static const uint16_t RTR_EXPIRATION_DEFAULT = 7200;
// ...
static const uint16_t RTR_REFRESH_DEFAULT = 3600;
// ...
static const uint16_t RTR_RETRY_DEFAULT = 600;
```

File: rtrlib/rtr_mgr.c [hunk 11]
```diff
@@ -431,17 +507,148 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+
+	struct rtr_mgr_group_node *gnode;
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
+		gnode = node->data;
+		if (gnode->group->preference == group->preference) {
+			RTR_DBG("Group with preference value already exists!");
+			return RTR_INVALID_PARAM;
+		}
```
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: Could there be a missing unlock before returning?

File: rtrlib/rtr_mgr.c (hunk 5 above)
Comment (mroethke) [inline rtrlib/rtr_mgr.c]: sockets_len is an unsigned int and so should j

File: rtrlib/rtr_mgr.c (hunk 3 above)
Comment (reuteran) [inline rtrlib/rtr_mgr.c]: Weird, this looks fine in my editor and also on github when I view the whole file. But in this view there is indeed a space missing...

File: rtrlib/rtr_mgr.c (hunk 4 above)
Comment (reuteran) [inline rtrlib/rtr_mgr.c]: Coding style script throws a warning because of 80 char line limit

File: rtrlib/rtr_mgr.c [hunk 12]
```diff
@@ -431,17 +525,156 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+		//TODO This is not pretty. It wants to get the same intervals
+		//that are being used by other groups. Maybe intervals should
+		//be store globally/per-group/per-socket?
+		if (gnode->group->sockets[0]->refresh_interval)
+			refresh_iv = gnode->group->sockets[0]->refresh_interval;
+		if (gnode->group->sockets[0]->retry_interval)
+			retry_iv = gnode->group->sockets[0]->retry_interval;
+		if (gnode->group->sockets[0]->expire_interval)
+			expire_iv = gnode->group->sockets[0]->expire_interval;
+		node = node->next;
```
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: I kind of forgot why we are assigning the intervals in the first place here (even though I think I wrote this).
Wouldn't it be better to leave the intervals at their defaults and add API functionality to set intervals? For now, this might only be for the whole group, e.g.:
`set_refresh_interval(struct rtr_mgr_group *group, unsigned int new_interval)`

File: rtrlib/rtr_mgr.c [hunk 13]
```diff
@@ -431,17 +525,156 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+			       new_group_node);
+	config->len++;
+
+	tommy_list_sort(&config->groups, &rtr_mgr_config_cmp);
+
+	struct rtr_mgr_group *best_group = rtr_mgr_get_first_group(config);
+
+	if (best_group->status == RTR_MGR_CLOSED) {
+		for (unsigned int j = 0; j < best_group->sockets_len; j++)
+			rtr_start(best_group->sockets[j]);
```
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: What is the behavior in case one or more sockets cannot be started?
Maybe do
```
if (rtr_start(best_group->sockets[j]) != RTR_SUCCESS)
    tr_free(best_group->sockets[j]);
```
or is this not even necessary?

File: rtrlib/rtr_mgr.c [hunk 14]
```diff
@@ -431,17 +525,152 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+	tommy_list_insert_tail(&config->groups, &new_group_node->node,
+			       new_group_node);
+	config->len++;
+
+	tommy_list_sort(&config->groups, &rtr_mgr_config_cmp);
+
+	struct rtr_mgr_group *best_group = rtr_mgr_get_first_group(config);
+
+	if (best_group->status == RTR_MGR_CLOSED) {
+		rtr_mgr_start_sockets(best_group);
```
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: Still not sure what should happen in case a socket is not started properly by `rtr_mgr_start_sockets`.

File: rtrlib/rtr_mgr.c (hunk 4 above)
Comment (fho) [inline rtrlib/rtr_mgr.c]: ah, we could remove the `unsigned` instead :-)

File: rtrlib/rtr_mgr.c (hunk 12 above)
Comment (reuteran) [inline rtrlib/rtr_mgr.c]: Yes, it would be better to be able to set intervals for each socket, since they are already stored by each socket. I think this should be in a separate pull request though.

**Comment by waehlisch** [comment]
To sum up, we have still two open questions:
 1.  How to handle socket specific timers? Suggestion by @reuteran is to put this in a separate PR.
 1. Which implementation for the link list? Suggestion is to postpone. We can beautify this later, it will introduce additionally required features.

@fho, can you give a final review?  We would like to get this merged soon, to progress with FRR ;).

File: tests/test_dynamic_groups.c [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	//remove 100 groups
+	for (int i = 0; i < 100; i++)
+		rtr_mgr_remove_group(conf, --group3.preference);
+
+	rtr_mgr_stop(conf);
+	rtr_mgr_free(conf);
+	free(group2.sockets);
+	free(groups[0].sockets);
```
Comment (nbars) [inline tests/test_dynamic_groups.c]: We are missing here a ```free(group3.sockets)``` I think.

File: rtrlib/rtr_mgr.c [hunk 16]
```diff
@@ -331,35 +379,41 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
...
-			MGR_DBG1("Error Empty sockets array in socket group!");
-			goto err;
-		}
-		for (unsigned int j = 0; j < cg.sockets_len; j++) {
-			if (rtr_init(cg.sockets[j], NULL,
+	/* Copy the groups from the array into linked list config->groups */
+	config->len = groups_len;
+	config->groups = NULL;
+	struct rtr_mgr_group *cg;
+	struct rtr_mgr_group_node *group_node;
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: variable declaration should only be done at the beginning of the block

File: rtrlib/rtr_mgr.c [hunk 17]
```diff
@@ -331,35 +379,41 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
...
+
+	for (unsigned int i = 0; i < groups_len; i++) {
+		cg = malloc(sizeof(struct rtr_mgr_group));
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: check missing if malloc failed

File: rtrlib/rtr_mgr.c [hunk 18]
```diff
@@ -331,35 +379,41 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
...
-			MGR_DBG1("Error Same preference for 2 socket groups!");
-			goto err;
-		}
-		if (cg.sockets_len == 0) {
-			MGR_DBG1("Error Empty sockets array in socket group!");
-			goto err;
-		}
-		for (unsigned int j = 0; j < cg.sockets_len; j++) {
-			if (rtr_init(cg.sockets[j], NULL,
+	/* Copy the groups from the array into linked list config->groups */
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: that part could be moved to a separate function to keep rtr_mgr_init() small and better readable

File: rtrlib/rtr_mgr.c [hunk 19]
```diff
@@ -331,35 +379,41 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
...
+		memcpy(cg, &groups[i], sizeof(struct rtr_mgr_group));
+
+		cg->status = RTR_MGR_CLOSED;
+		for (unsigned int j = 0; j < cg->sockets_len; j++) {
+			if (rtr_init(cg->sockets[j], NULL,
 				     pfxt, spki_table, refresh_interval,
 				     expire_interval, retry_interval,
-				     rtr_mgr_cb, config) != RTR_SUCCESS) {
+				     rtr_mgr_cb, config, cg) != RTR_SUCCESS) {
 				err_code = RTR_INVALID_PARAM;
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: instead of assigning err_code to the (currently) only error code that rtr_init() can return it would be better to do something like:
```
err = rtr_init();
if (err)
    goto [..]
```
If rtr_init() would change in the future to also return other errors, it's easy to miss to update this place here to ensure that the user gets the right error code back :-)

File: rtrlib/rtr_mgr.c [hunk 20]
```diff
@@ -331,35 +379,41 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
...
 				goto err;
 			}
 		}
-		last_preference = cg.preference;
+		group_node = malloc(sizeof(struct rtr_mgr_group_node));
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: check missing if malloc() failed

File: rtrlib/rtr_mgr.c [hunk 21]
```diff
@@ -331,35 +379,41 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
...
+		group_node->group = cg;
+		tommy_list_insert_tail(&config->groups, &group_node->node,
+				       group_node);
 	}
+	/* Our linked list should be sorted already, since the groups array was
+	 * sorted. However, for safety reasons we sort again.
+	 */
+	tommy_list_sort(&config->groups, &rtr_mgr_config_cmp);
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: that's unnecessary, it **must** be sorted already :-)

File: rtrlib/rtr_mgr.c (hunk 17 above)
Comment (fho) [inline rtrlib/rtr_mgr.c]: The allocated memory seems to be leaked if the function returns because rtr_init() failed

File: rtrlib/rtr_mgr.c [hunk 22]
```diff
@@ -377,37 +431,77 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
 	return err_code;
 }
 
+static struct rtr_mgr_group *rtr_mgr_get_first_group(struct rtr_mgr_config
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: name could be shortened to `rtr_mgr_first_group`

File: rtrlib/rtr_mgr.c [hunk 23]
```diff
@@ -377,37 +431,77 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
...
+						     *config)
+{
+	tommy_node *head = tommy_list_head(&config->groups);
+	struct rtr_mgr_group_node *group_node = head->data;
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: the assignment could be removed and return head->data->group directly

File: rtrlib/rtr_mgr.c [hunk 24]
```diff
@@ -377,37 +431,77 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
 	return err_code;
 }
 
+static struct rtr_mgr_group *rtr_mgr_get_first_group(struct rtr_mgr_config
+						     *config)
+{
+	tommy_node *head = tommy_list_head(&config->groups);
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: the function doesn't need to acquire the mutex?

File: rtrlib/rtr_mgr.c [hunk 25]
```diff
@@ -431,17 +525,155 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+int rtr_mgr_add_group(struct rtr_mgr_config *config,
+		      const struct rtr_mgr_group *group)
+{
+	unsigned int refresh_iv = 3600;
+	unsigned int retry_iv = 600;
+	unsigned int expire_iv = 7200;
+
+	pthread_mutex_lock(&config->mutex);
+
+	struct rtr_mgr_group_node *gnode;
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: declaration should be done before statements

File: rtrlib/rtr_mgr.c [hunk 26]
```diff
@@ -431,17 +525,155 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
+		gnode = node->data;
+		if (gnode->group->preference == group->preference) {
+			MGR_DBG1("Group with preference value already exists!");
+			pthread_mutex_unlock(&config->mutex);
+			return RTR_INVALID_PARAM;
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: instead of having to enable debugging to find out the  cause it would be great to return an error code that describes the reason more specific

File: rtrlib/rtr_mgr.c [hunk 27]
```diff
@@ -431,17 +525,155 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+		}
+	}
+
+	struct rtr_mgr_group_node *new_group_node = malloc(
+					sizeof(struct rtr_mgr_group_node));
+	if (!new_group_node) {
+		pthread_mutex_unlock(&config->mutex);
+		free(new_group);
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: use a goto error handler to prevent duplicating code in case of an error

File: rtrlib/rtr_mgr.c [hunk 28]
```diff
@@ -431,17 +525,155 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+	struct rtr_mgr_group *new_group = malloc(sizeof(struct rtr_mgr_group));
+
+	if (!new_group) {
+		pthread_mutex_unlock(&config->mutex);
+		return RTR_ERROR;
+	}
+	memcpy(new_group, group, sizeof(struct rtr_mgr_group));
+	new_group->status = RTR_MGR_CLOSED;
+
+	for (unsigned int j = 0; j < new_group->sockets_len; j++) {
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: maybe that can be outsourced to a separate function and the same code can be used from rtr_mgr_init() and rtr_mgr_add_group()

File: rtrlib/rtr_mgr.c [hunk 29]
```diff
@@ -431,17 +525,155 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
...
+int rtr_mgr_remove_group(struct rtr_mgr_config *config,
+			 unsigned int preference)
+{
+	pthread_mutex_lock(&config->mutex);
+	tommy_node *remove_node = NULL;
+	tommy_node *node = tommy_list_head(&config->groups);
+	struct rtr_mgr_group_node *group_node;
+	struct rtr_mgr_group *remove_group;
+
+	// Find the node of the group we want to remove
```
Comment (fho) [inline rtrlib/rtr_mgr.c]: could be a separate function _find_node() or _remove_node() (which also finds it)

File: rtrlib/rtr_mgr.h [hunk 30]
```diff
@@ -67,28 +68,37 @@ struct rtr_mgr_group {
 	enum rtr_mgr_status status;
 };
 
+//TODO Find a nicer way todo a linked list (without writing our own)
+struct rtr_mgr_group_node {
```
Comment (fho) [inline rtrlib/rtr_mgr.h]: Seems that the `struct rtr_mgr_group_node` is only used internally in rtr_mgr.c if it's true it should be moved to the .c file

**Comment by fho** [comment]
> To sum up, we have still two open questions:
> 
> How to handle socket specific timers? Suggestion by @reuteran is to put this in a separate PR.

sounds good to me

> Which implementation for the link list? Suggestion is to postpone. We can beautify this later, it will introduce additionally required features.

Seems that the list is only used internally on contrary what I first thought. :-)
I agree, which linked list implementation is used doesn't really matter here.

> @fho, can you give a final review? We would like to get this merged soon, to progress with FRR ;).

Yes, sorry, I was spontaneously on a weekend trip and didn't manage to do it like planned.
I added more comments but I still wasn't able to go through all of the code neither to understand all the logic.
I'll try to have another look soon, if it wasn't merged then already. :-)
It looks good in general, cool improvement!

File: rtrlib/rtr_mgr.h (hunk 30 above)
Comment (ColinBS) [inline rtrlib/rtr_mgr.h]: The struct is also used in the `test_dynamic_groups` test. So we keep it in the header file?

File: rtrlib/rtr_mgr.c (hunk 23 above)
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: For some reason this won't work. No idea why not, though...

File: rtrlib/rtr_mgr.c (hunk 18 above)
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: I will try to make this a PR on it's own. @mroethke suggested to use the `add_group` function but there might be some refactoring necessary to make this happen.

File: rtrlib/rtr_mgr.c (hunk 22 above)
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: This might also be changed in a separate PR that does some refactoring. In my humble opinion, I think the name is OK, the _get_ prevents confusion over what exactly the function does, I think.

File: rtrlib/rtr_mgr.c (hunk 21 above)
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: Would like to ask @reuteran and @mroethke about this.

File: rtrlib/rtr_mgr.c (hunk 24 above)
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: Not sure. @reuteran and @mroethke ?

File: rtrlib/rtr_mgr.c (hunk 29 above)
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: Could be part of a separate PR.

File: rtrlib/rtr_mgr.c (hunk 6 above)
Comment (fho) [inline rtrlib/rtr_mgr.c]: I meant the `err_code` variable :

File: rtrlib/rtr_mgr.h (hunk 30 above)
Comment (fho) [inline rtrlib/rtr_mgr.h]: No, the external API should be minimal and only contain declaration that have a purpose during normal use of the library.
Either the test should use only the external library or we have to find another way to test it. (including the .c file in the test might work)

Comment (mroethke) [inline rtrlib/rtr_mgr.h]: We need a way to access the the group list tough or at least access the active group, otherwise it would not be possible to get status information beyond "it still works somehow".

Of course that can also be archived with helper functions.

File: rtrlib/rtr_mgr.c (hunk 6 above)
Comment (ColinBS) [inline rtrlib/rtr_mgr.c]: Ohhhh... well... huh...


Will do that!

**Comment by mroethke** [comment]
I just re based this onto master and made the necessary changes to properly use the alloc utils. 

I furthermore made a change to how socket initialization handles the configuration. While not strictly related to dynamic groups it fixes an error that could could lead to memory leaks that is basically irrelevant without this patch.

File: rtrlib/rtr_mgr.c (hunk 5 above)
Comment (smlng) [inline rtrlib/rtr_mgr.c]: why isn't `sockets_len` of type `size_t` anyway? Or alternatively just `unsigned` should suffice, too.

File: rtrlib/rtr_mgr.c [hunk 31]
```diff
@@ -91,135 +120,171 @@ bool rtr_mgr_config_status_is_synced(const struct rtr_mgr_group *group)
...
-		    (config->groups[i].status == RTR_MGR_CLOSED))
-			return &config->groups[i];
+	struct rtr_mgr_group_node *group_node;
+	struct rtr_mgr_group *group;
+
+	pthread_mutex_lock(&config->mutex);
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
+	group_node = node->data;
```
Comment (smlng) [inline rtrlib/rtr_mgr.c]: indention looks of by one tab here

File: rtrlib/rtr_mgr.c [hunk 32]
```diff
@@ -91,135 +120,171 @@ bool rtr_mgr_config_status_is_synced(const struct rtr_mgr_group *group)
...
 static bool is_some_rtr_mgr_group_established(struct rtr_mgr_config *config)
 {
-	for (int i = 0; i < config->len; i++) {
-		if (config->groups[i].status == RTR_MGR_ESTABLISHED)
+	struct rtr_mgr_group_node *group_node;
+
+	pthread_mutex_lock(&config->mutex);
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
```
Comment (smlng) [inline rtrlib/rtr_mgr.c]: a for-loop would fit here perfectly, or not?

```
for (tommy_node *node = tommy_list_head(&config->groups); 
     node != NULL; node = node->next) {
...
}
```

File: rtrlib/rtr_mgr.c [hunk 33]
```diff
@@ -91,135 +120,171 @@ bool rtr_mgr_config_status_is_synced(const struct rtr_mgr_group *group)
...
-		set_status(config, &config->groups[ind],
-			   config->groups[ind].status, sock);
+		set_status(config, group,
+			   group->status, sock);
 }
 
 static inline void _rtr_mgr_cb_state_established(const struct rtr_socket *sock,
 						 struct rtr_mgr_config *config,
-						 unsigned int ind)
+						 struct rtr_mgr_group *my_group)
```
Comment (smlng) [inline rtrlib/rtr_mgr.c]: inconsistent parameter naming, above its just `group` which is fine - `my_group` looks a bit like code from a beginners tutorial 😉

File: rtrlib/rtr_mgr.c [hunk 34]
```diff
@@ -91,135 +120,171 @@ bool rtr_mgr_config_status_is_synced(const struct rtr_mgr_group *group)
...
-			    cg.status != RTR_MGR_ERROR &&
-			    cg.status != RTR_MGR_CLOSED &&
-			    cg.preference < config->groups[ind].preference)
+		struct rtr_mgr_group_node *group_node;
+		struct rtr_mgr_group *current_group;
+
+		pthread_mutex_lock(&config->mutex);
+		tommy_node *node = tommy_list_head(&config->groups);
+
+		while (node) {
```
Comment (smlng) [inline rtrlib/rtr_mgr.c]: again for-loop might fit better here, see above

File: rtrlib/rtr_mgr.c [hunk 35]
```diff
@@ -91,135 +120,171 @@ bool rtr_mgr_config_status_is_synced(const struct rtr_mgr_group *group)
...
+			group_node = node->data;
+			current_group = group_node->group;
+			if (current_group != my_group &&
+			    current_group->status != RTR_MGR_ERROR &&
```
Comment (smlng) [inline rtrlib/rtr_mgr.c]: if would put parentheses around each condition

File: rtrlib/rtr_mgr.c [hunk 36]
```diff
@@ -371,44 +449,87 @@ int rtr_mgr_init(struct rtr_mgr_config **config_out,
...
+	bool all_sync;
+	struct rtr_mgr_group_node *group_node;
 
-		for (unsigned int j = 0;
-		     all_sync && (j < config->groups[i].sockets_len); j++) {
-			if (config->groups[i].sockets[j]->last_update == 0)
+	pthread_mutex_lock(&config->mutex);
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
```
Comment (smlng) [inline rtrlib/rtr_mgr.c]: use for-loop? See previous comments ...

File: rtrlib/rtr_mgr.c (hunk 32 above)
Comment (colinbs) [inline rtrlib/rtr_mgr.c]: In my opinion, the readability suffers from this more compact solution.

File: rtrlib/lib/alloc_utils.h [hunk 37]
```diff
@@ -34,4 +34,13 @@ void lrtr_free(void *ptr);
 
 void *lrtr_realloc(void *ptr, size_t size);
 
+/**
+ * @brief Duplicates a string
+ * @param[in] string
+ * @returns Duplicated string
+ * @returns NULL if string is NULL
+ * @returns NULL on error
```
Comment (mroethke) [inline rtrlib/lib/alloc_utils.h]: The main and currently only use case for this function is copying the socket configuration, parts of which may be null. Instead of guarding every copy I decided to write the function accordingly.

Comment (mroethke) [inline rtrlib/lib/alloc_utils.h]: You are right.
I changed to function to match std strdup and added checks where necessary.

File: rtrlib/rtr_mgr.c (hunk 32 above)
Comment (smlng) [inline rtrlib/rtr_mgr.c]: fine by me, leave as is.

**Comment by smlng** [comment]
@mroethke and @colinbs please consider #132
